File: 프로그래머스/2/87946. 피로도/피로도.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

int maxDungeons = 0;

void backtrack(int k, int count, const vector<vector<int>>& dungeons, vector<bool>& visited) {
    maxDungeons = max(maxDungeons, count); // 최대 탐험 던전 수 갱신

    for (int i = 0; i < dungeons.size(); ++i) {
        if (!visited[i] && k >= dungeons[i][0]) { // 던전 탐험 가능 여부 확인
            visited[i] = true; // 현재 던전 방문 처리
            backtrack(k - dungeons[i][1], count + 1, dungeons, visited); // 재귀 호출
            visited[i] = false; // 백트래킹: 방문 처리 해제
        }
    }
}

int solution(int k, vector<vector<int>> dungeons) {
    vector<bool> visited(dungeons.size(), false);
    backtrack(k, 0, dungeons, visited);
    return maxDungeons;
}

File: 프로그래머스/2/87946. 피로도/피로도.cs
using System;
using System.Collections.Generic;
using System.Linq;

public class Solution 
{
    bool[] visited;
    int[] order;
    List<int> result;
    public int solution(int k, int[,] dungeons) 
    {
        int length = dungeons.GetLength(0);
        visited = new bool[length];
        order = new int[length];
        result = new List<int>();
        Search(0, k, length, dungeons);
        return result.ToArray().Max();
    }
    
    void Search(int depth, int k, int length, int[,] dungeons)
    {
        if (depth == length)
        {
            int count = 0;
            int remain = k;
            for (int i = 0; i < order.Length; i++)
            {
                int index = order[i];
                
                if(remain >= dungeons[index, 0])
                {
                    remain -= dungeons[index, 1];
                    count++;
                }
                else
                {
                    break;
                }
                
            }
            result.Add(count);
            Console.WriteLine();
            return;
        }

        for (int i = 0; i < length; i++)
        {
            if (visited[i])
                continue;

            visited[i] = true;
            order[depth] = i;

            Search(depth + 1, k, length , dungeons);
            visited[i] = false;
        }
    }
}
